### PR TITLE
Add SharedBlobCacheService.maybeFetchRange method

### PR DESCRIPTION
**Pull request by tlrx**
This change adds a `maybeFetchRange()` method that can be used to fetch a given range in a specific blob region if at least a free page is available. The range can be larger than the blob region and is remapped (`mapSubRangeToRegion();`) before hitting the cache.

It is basically a variant of `maybeFetchRegion` that allow passing a range instead of the whole region.

### CODE REVIEW

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 1]
```diff
@@ -568,6 +568,73 @@ public void maybeFetchRegion(
...
+     * <p>
+     * This method returns as soon as the download tasks are instantiated, but the tasks themselves
+     * are run on the bulk executor.
+     * <p>
+     * If an exception is thrown from the writer then the cache entry being downloaded is freed
+     * and unlinked
+     *
+     * @param cacheKey      the key to fetch data for
+     * @param region        the region of the blob
+     * @param range         the range in the region to fetch
```
Comment (henningandersen) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:582]: I think this should be "the range in the blob to fetch"? At least the wording here made me think it was relative to the region start.

**Review by henningandersen — COMMENTED** [review]
Looks good. It would be nice to add a test of this too, similar to `testMaybeFetchRegion` maybe?

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java]: Right, I pushed 970394895889a908029084156569704e4b207dc5.

I must admit I took a shortcut here by allowing to pass a range of the blob that cannot be included within the `region`, letting the method implementation to remap the range to the region and call the listener on empty range.

I think we do this in other places, not great, but it simplifies the calling methods.

**Comment by tlrx** [comment]
Thanks Henning for the feedback. I added tests in 38bc83e80807c482110830f944c4db47e3820688

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by tlrx** [comment]
That was quick, thanks Henning!

**Pull request merged**